Upgrade CallbackFunction default runtime to Node 20.x

t0yv0 opened the pull request:
This is a breaking change for users of Node automatically provisioned lambdas, moving them from the Node 16.x to Node 20.x runtime without an ability to opt out.
 
Prior to this change, Lambda functions provisioned in Node with CallbackFunction and helper methods such as Bucket.onObjectRemoved, Bucket.onObjectCreated or sns.Topic.onEvent used Node 16.x runtime by default. Per [lambda-runtimes.html](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtimes-supported) this is now deprecated by AWS. After the change, the new default runtime is Node 20.x.

Fixes #4134 

The change is supported by strengthening a test to utilize `fetch()` API that was made available in 18.x and exercise the callback function at runtime.

corymhall commented on sdk/nodejs/lambda/lambdaMixins.ts:
```diff
@@ -406,13 +406,11 @@ export class CallbackFunction<E, R> extends LambdaFunction {
             code: code,
             handler: serializedFileNameNoExtension + "." + handlerName,
-            runtime: args.runtime || Runtime.NodeJS16dX,
+            runtime: args.runtime || Runtime.NodeJS18dX,
```
Should we take the opportunity to just upgrade to 20x?

t0yv0 replied:
You are right, https://github.com/pulumi/pulumi-aws/issues/4134 was suggesting 18 as the current LTS but https://endoflife.date/nodejs  suggests 20 is the current LTS so that should be reasonable.

corymhall reviewed (approved):
Just 1 suggestion

flostadler reviewed (changes requested):
Even though it's a breaking change I think not doing it would be a greater issue for our users.
Function creates using node 16 will be blocked on Feb 28, 2025 and updates will be blocked on Mar 31, 2025.

We should really directly upgrade to node 20 though. The node 18 runtime is still using AL2 under the hood (see [AWS docs](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html#runtimes-supported)) which will be deprecated in June 2025. There's a high chance they'll deprecate Node 18 as well in the near future.

flostadler commented:
We should mark the Node 16.x runtime as deprecated as well: https://github.com/pulumi/pulumi-aws/blob/master/provider/resources.go#L4541

flostadler commented on sdk/nodejs/lambda/lambdaMixins.ts:
```diff
@@ -406,13 +406,11 @@ export class CallbackFunction<E, R> extends LambdaFunction {
         // Copy over all option values into the function args.  Then overwrite anything we care
```
Can't put the comment on that line because it's outside of the PR, but we should also update the docs to reflect the new default: https://github.com/pulumi/pulumi-aws/blob/7461d6834b90936a38e80cc4bac2c10ff97499a0/sdk/nodejs/lambda/lambdaMixins.ts#L227

flostadler reviewed (approved):
Nice!

t0yv0 commented:
PTAL @flostadler addressed the feedback.

> Even though it's a breaking change I think not doing it would be a greater issue for our users.

I suspect just a slightly unexpected upgrade. Just calling it out for the release notes, still doing a minor release on this.

mikhailshilkov commented:
This PR has been shipped in [release v6.49.0](https://github.com/pulumi/pulumi-aws/releases/tag/v6.49.0).
